compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1406,6 +1406,63 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         true
     }
 
+    #[instrument(skip(self, err))]
+    pub(crate) fn suggest_unit_or_empty_struct_construction(
+        &self,
+        err: &mut Diagnostic,
+        expr: &hir::Expr<'_>,
+        expected_ty: Ty<'tcx>,
+    ) -> bool {
+        // It's useless to suggest a fix if it's from a foreign macro
+        if in_external_macro(self.tcx.sess, expr.span) {
+            return false;
+        }
+
+        // Expected type needs to be an ADT
+        let Some(adt_def) = expected_ty.ty_adt_def() else {
+            return false;
+        };
+
+        // ADT needs to be a struct and have no fields
+        if !adt_def.is_struct() || !adt_def.is_payloadfree() {
+            return false;
+        }
+
+        let mut ctors = adt_def.ctors();
+        let first_ctor_kind = ctors.next();
+
+        // If we have multiple ctors, then this is not a struct
+        //
+        // This shouldn't occur (since we check that this is a struct earlier)
+        // but we can safely handle this case anyways.
+        if ctors.next().is_some() {
+            return false;
+        }
+
+        // Get the name of the expected type
+        let mut type_name = with_no_trimmed_paths!(self.tcx.def_path_str(adt_def.did()));
```

Comment:
I don't think we'll run into this? If the path isn't imported, we hit an "cannot find type in this scope". This doesn't trigger this new diags because at this point we have the expected type, right?